packages/jsii-diff/README.md

```diff
@@ -0,0 +1,82 @@
+# jsii-diff
+
+__jsii-diff__ compares two jsii assemblies for compatibility.
+
+In the future, it will be able to do generic comparisons, but for
+now it will compare assemblies for API compatibility, and exit
+with a non-zero exit code if any **stable** APIs have had incompatible
+changes.
+
+API items that have no stability are treated as **experimental** (i.e.,
+do not affect API compatibility.
+
+## Usage
+
+To compare two JSII packages:
+
+    jsii-diff <path/to/old> [path/to/new]
+
+To compare current package against latest published NPM release:
+
+    # Can leave out . for current directory
+    jsii-diff npm:package [.]
+
+## Details
+
+__jsii-diff__ will assert that code written against version **A** of a library
+will still typecheck when compiled against version **B** of that library. It
+does this by verifying the following properties:
+
+- Any type (class/interface/enum) in **A** must also exist in **B**.
+- Enums have only added members.
+- Classes and behavioral interfaces have only added members, restricted
+  property types and method return types, and only widened argument types.
+
+### Subclassing
+
+In principle, in order to stay API compatible no members can ever be added to
+interfaces, and no abstract members can ever be added to classes. This is
+because any user of the library could be implementing the interface or could
+be inheriting frmo the class. In those cases, the addition of an abstract
+member would break their code, since they wouldn't be implementing that
+member yet.
+
+Because this contract is a little too restrictive for evolution of libraries,
+we assume most interfaces will only be implemented by library classes, so we
+can disregard burden on implementors as part of API compatibility evaluation.
+
+If an interface or class is intended to be subclassed by library users,
+annotate it with the `@subclassable` annotation. In that case, we will make
+sure no (abstract) members are added to it.
+
+### Struct types
+
+Structs (data types, interfaces consisting completely of `readonly`
```

Comment:
What's the advantage of the annotation over the analysis that's currently being done already?

Reply:
I am concerned about the case where some enterprising engineer will decide to return an input struct as a return value in V1 and then it will be impossible to extend this struct from this point forward. Being explicit about a struct being an output type will allow us to enforce that input structs and output structs cannot be mixed up.

Reply:
It'll never be *completely* impossible to extend. For both input and output structs it's allowed to add optional fields. And if we mess up completely, there's always `@deprecated`.

I feel hesitant to add ever more annotation burdens on users, especially since I don't know that it's going to add value, or if this is going to be a real problem.

The same enterprising engineer, when faced with the error `Cannot use this struct as output unless you annotate it with @outputvalue`, is going to put that annotation there and continue with their life.

Reply:
> Plus, that same enterprising engineer, when faced with the error Cannot use this struct as output unless you annotate it with @outputvalue, is going to put that annotation there and continue with their life.

They won't be able to, because they will get an error `cannot use an @output struct as a method argument type`.

I am okay if we punt this for another iteration, but I do feel like this has the potential to be problematic if we leave it implicit.

Reply:
In that case we can do that same analysis without annotating, disallowing the same type used in both input and output position.

But doing so would immediately clash with our reasonable solution to make L1's fake-mutable:

```ts
interface BucketProps { ... }

class Bucket {
  constructor(scope: Construct, id: string, props: BucketProps) { ... }

  public overrideProperty(name: string, value: any) { ... }
  
  // Access current properties
  // DISALLOWED, can't use BucketProps both as input and output
  public get properties(): BucketProps { ... }
}
```

I would also prefer to punt and see if it actually turns out to be a problem.

Reply:
+1 to punt.

BTW, I actually think that this brings up an issue with this `get properties()` solution (which we still haven't implemented). `BucketProps` is only shallowly readonly, which means that people will be able to modify some deep structure within the props tree but not the root?! That's messy. I rather we generate another type `BucketPropOverrides` or something like that which will be fully read/write for this use case (and then we are back in nice @output-only land).

Reply:
> `BucketProps` is only shallowly readonly

It is not, really, since all deep members must also be structs, hence will also have only `readonly` members.

The only case in which this is not true is for arrays, because we don't want to use the `ReadonlyArray<>` type but instead prefer to wait until TypeScript has a `deepreadonly` keyword. Oh, and I suppose objects-that-are-actually-maps. 

So I'd prefer to think about it as "we do have deep immutability except it's not perfectly advertised in the type system yet".